### ISSUE TITLE
Global ID returned in Query vs. ID used in Mutation

### ISSUE THREAD

**Issue by vangavroche** [RisingStack/graffiti-mongoose#70]
It looks like the object ID returned in query is converted to GlobalId (by the graphql-relay since the globalIdField is used).

But in the mutation, the ID sent from the client is directly passed to the Mongo so some casting error will be reported. https://github.com/RisingStack/graffiti-mongoose/blob/master/src/schema/schema.js#L120

Will it be fixed in the framework, or should I handle it in the hooks?

Thanks

**Comment by tothandras** [comment]
Hi @vangavroche! 
Did you try it? Mutations should handle both global (`id`) and mongoose (`_id`) ids.

**Comment by vangavroche** [comment]
In the example https://github.com/RisingStack/graffiti-mongoose/blob/master/example/user.js where User has friends that are also "User"s

After I created two users (without any friends), the query responses of "users {id, name}" are like

```
{
  "data": {
    "users": [
      {
        "id": "VXNlcjo1NjY1YWZkMGU1ZWQwMTAzMWU5ODc3MjQ=",
        "name": "Kaka"
      },
      {
        "id": "VXNlcjo1NjY2NWI0M2U1ZWQwMTAzMWU5ODc3MjY=",
        "name": "Messi"
      }
    ]
  }
}
```

The generated mutation to add user has input that contains 
```
name: String
age: Float
createdAt: Date
friends: [ID]
clientMutationId: String!
```

When I try to add a user using mutation

```
mutation M {
  addUser(input: {name:"Ronaldo", age: 12, friends:["VXNlcjo1NjY2NWI0M2U1ZWQwMTAzMWU5ODc3MjY="], clientMutationId:"1234"}) {
    changedUserEdge {
      node {
        name
      }
    }
  }
}
```

It won't work, since the ID is not supposed a valid one.

**Comment by vangavroche** [comment]
Do you mean when I query the ID to set such association ("friends" in this case), I should explicitly get the _id to use?

Using query like this:
```
query U {
  users {
    _id, name
  }
}
```

**Comment by vangavroche** [comment]
BTW, is there any special reason why the singular object reference is not supported in the mutation field generation?

https://github.com/RisingStack/graffiti-mongoose/blob/master/src/schema/schema.js#L124

In the case like 
```
//
const TodoSchema = new mongoose.Schema({
  text: {
    type: String
  },
  createdAt: {
    type: Date,
    default: Date.now()
  },
  completed: {
    type: Boolean,
    default: false
  },
  owner: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }
});
```

 owner field is not used as the mutation input.

To me it seems an easy problem to solve

```
else {
        // In this case it is assume that the field is a singular object reference like
        // {
        //    type: ObjectId,
        //    ref: "User"
        // }
        //
        inputFields[field.name] = {
          name: field.name,
          type: _graphql.GraphQLID
        };
      }
```


But I wonder if there's anything I'm missing.

**Comment by tothandras** [comment]
@vangavroche Okay, the add mutation error is a real bug, see #71 for a quick fix.
I am looking into the object input now.

**Comment by tothandras** [comment]
@vangavroche I've added the reference support too, although we might need more testing.

**Comment by vangavroche** [comment]
@tothandras  thanks.

**Comment by tothandras** [comment]
@vangavroche Can you verify that it resolves your issues? :) I could have missed something. Please add a thumbs up to the PR, I will merge it then.

**Comment by vangavroche** [comment]
Yes, it works and solves the issue.

**Comment by tothandras** [comment]
Released under `v4.3.2`.
